### PR TITLE
Code Review 8/29/18

### PR DESCRIPTION
**Pull request by jtcramer**

### CODE REVIEW

File: l3embedding/train.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    }
+    LOGGER.info('Training with the following arguments: {}'.format(param_dict))
+
+    if continue_model_dir:
+        latest_model_path = os.path.join(continue_model_dir, 'model_latest.h5')
+        m, inputs, outputs = load_model(latest_model_path, model_type, return_io=True, src_num_gpus=gpus)
+    else:
+        m, inputs, outputs = MODELS[model_type](num_gpus=gpus)
+
+    loss = 'binary_crossentropy'
```
Comment (jtcramer) [inline l3embedding/train.py]: The loss is the same as as categorical_crossentropy, up to a constant. But binary_crossentropy = 2 * categorical_crossentropy. This shouldn't cause any serious optimization issues, but effectively halves the weight decay constant.

File: l3embedding/audio_model.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+    asr = 48000
+    audio_window_dur = 1
+    # INPUT
+    x_a = Input(shape=(1, asr * audio_window_dur), dtype='float32')
+
+    # SPECTROGRAM PREPROCESSING
+    # 257 x 199 x 1
+    y_a = Spectrogram(n_dft=n_dft, n_hop=n_hop, power_spectrogram=1.0, # n_win=n_win,
+                      return_decibel_spectrogram=True, padding='valid')(x_a)
+    y_a = BatchNormalization()(y_a)
```
[the remark this answers is not in the document]
Comment (jtcramer) [reply l3embedding/audio_model.py]: Or leave it in, depending on what we decide

File: l3embedding/vision_model.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    outputs: Model outputs
+            (Type: keras.layers.Layer)
+    """
+    weight_decay = 1e-5
+    ####
+    # Image subnetwork
+    ####
+    # INPUT
+    x_i = Input(shape=(224, 224, 3), dtype='float32')
+    y_i = BatchNormalization()(x_i)
```
[the remark this answers is not in the document]
Comment (jtcramer) [reply l3embedding/vision_model.py]: Or leave it in, depending on what we decide.

File: classifier/train.py [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+    m.fit(x=X_train, y=y_train, batch_size=batch_size, epochs=num_epochs,
+          validation_data=validation_data, validation_split=valid_split,
+          callbacks=cb, verbose=2)
+
+    # Compute metrics for train and valid
+    train_pred = m.predict(X_train)
+    train_metrics = compute_metrics(y_train, train_pred, num_classes=num_classes)
+    # Set up train and validation metrics
+    train_metrics = {
+        'loss': metric_cb.train_loss[-1],
```
[the remark this answers is not in the document]
Comment (jtcramer) [reply classifier/train.py]: Also, we need to make sure we're reloading the checkpoint with the best loss, since Keras just leaves the model in whatever state the model was when finished training

File: classifier/metrics.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    Returns:
+        metrics: Metrics dictionary
+                 (Type: dict[str, *])
+    """
+    # Convert from one-hot to integer encoding if necessary
+    if y.ndim == 2:
+        y = np.argmax(y, axis=1)
+    if pred.ndim == 2:
+        pred = np.argmax(pred, axis=1)
```
[the remark this answers is not in the document]
Comment (jtcramer) [reply classifier/metrics.py]: This shouldn't actually have affected anything since we're calling `arr.ndim`, but it's a good precaution
